### Target

**Find the articles which are judged as Xinhua net but not is Xinhua net. **

In [1]:
import pandas as pd

In [2]:
import numpy as np

In [3]:
import matplotlib.pyplot as plt

In [4]:
import jieba

In [5]:
def cut(string): return ' '.join(jieba.cut(string))

In [22]:
content = pd.read_csv('../dataset/export_sql_1558435.zip', compression='zip', encoding='gb18030')

In [14]:
content.head()

id      author                  source  \
0  89617         NaN  快科技@http://www.kkj.cn/   
1  89616         NaN  快科技@http://www.kkj.cn/   
2  89615         NaN  快科技@http://www.kkj.cn/   
3  89614         NaN                     新华社   
4  89613  胡淑丽_MN7479                   深圳大件事   

                                             content  \
0  此外，自本周（6月12日）起，除小米手机6等15款机型外，其余机型已暂停更新发布（含开发版/...   
1  骁龙835作为唯一通过Windows 10桌面平台认证的ARM处理器，高通强调，不会因为只考...   
2  此前的一加3T搭载的是3400mAh电池，DashCharge快充规格为5V/4A。\r\n...   
3    这是6月18日在葡萄牙中部大佩德罗冈地区拍摄的被森林大火烧毁的汽车。新华社记者张立云摄\r\n   
4  （原标题：44岁女子跑深圳约会网友被拒，暴雨中裸身奔走……）\r\n@深圳交警微博称：昨日清...   

                                             feature  \
0  {"type":"科技","site":"cnbeta","commentNum":"37"...   
1  {"type":"科技","site":"cnbeta","commentNum":"15"...   
2  {"type":"科技","site":"cnbeta","commentNum":"18"...   
3  {"type":"国际新闻","site":"环球","commentNum":"0","j...   
4  {"type":"新闻","site":"网易热门","commentNum":"978",...   

                           title  \
0           小米MIUI 9首批机型曝光：共计15款   
1     骁龙835在Windows 10上的性能表现有望改善   
2      一加手机5细节曝光：3300mAh、充半小时用1天   
3  葡森林火灾造成至少62人死亡 政府宣布进入紧急状态（组图）   
4       44岁女子约网友被拒暴雨中裸奔 交警为其披衣相随   

                                                 url  
0     http://www.cnbeta.com/articles/tech/623597.htm  
1     http://www.cnbeta.com/articles/tech/623599.htm  
2     http://www.cnbeta.com/articles/tech/623601.htm  
3  http://world.huanqiu.com/hot/2017-06/10866126....  
4  http://news.163.com/17/0618/00/CN617P3Q0001875...

In [23]:
content['content'] = content['content'].fillna('')
content['content'] = content['content'].apply(cut)

In [24]:
content['title'] = content['title'].fillna('')
content['title'] = content['title'].apply(cut)

In [25]:
content.head()

id      author                  source  \
0  89617         NaN  快科技@http://www.kkj.cn/   
1  89616         NaN  快科技@http://www.kkj.cn/   
2  89615         NaN  快科技@http://www.kkj.cn/   
3  89614         NaN                     新华社   
4  89613  胡淑丽_MN7479                   深圳大件事   

                                             content  \
0  此外 ， 自 本周 （ 6 月 12 日 ） 起 ， 除 小米 手机 6 等 15 款 机型...   
1  骁龙 835 作为 唯一 通过 Windows   10 桌面 平台 认证 的 ARM 处理...   
2  此前 的 一加 3T 搭载 的 是 3400mAh 电池 ， DashCharge 快充 规...   
3  这是 6 月 18 日 在 葡萄牙 中部 大 佩德罗 冈 地区 拍摄 的 被 森林 大火 烧...   
4  （ 原 标题 ： 44 岁 女子 跑 深圳 约会 网友 被 拒 ， 暴雨 中 裸身 奔走 …...   

                                             feature  \
0  {"type":"科技","site":"cnbeta","commentNum":"37"...   
1  {"type":"科技","site":"cnbeta","commentNum":"15"...   
2  {"type":"科技","site":"cnbeta","commentNum":"18"...   
3  {"type":"国际新闻","site":"环球","commentNum":"0","j...   
4  {"type":"新闻","site":"网易热门","commentNum":"978",...   

                                          title  \
0                小米 MIUI   9 首批 机型 曝光 ： 共计 15 款   
1         骁龙 835 在 Windows   10 上 的 性能 表现 有望 改善   
2         一加 手机 5 细节 曝光 ： 3300mAh 、 充 半小时 用 1 天   
3  葡 森林 火灾 造成 至少 62 人 死亡   政府 宣布 进入 紧急状态 （ 组图 ）   
4       44 岁 女子 约 网友 被 拒 暴雨 中 裸奔   交警 为 其 披衣 相随   

                                                 url  
0     http://www.cnbeta.com/articles/tech/623597.htm  
1     http://www.cnbeta.com/articles/tech/623599.htm  
2     http://www.cnbeta.com/articles/tech/623601.htm  
3  http://world.huanqiu.com/hot/2017-06/10866126....  
4  http://news.163.com/17/0618/00/CN617P3Q0001875...

In [26]:
xinhuashe_news = content[content['source'].str.contains('新华', na=False)]

In [27]:
len(xinhuashe_news)

78855

In [28]:
len(content)

89611

In [29]:
len(xinhuashe_news) / len(content)

0.8799700929573379

In [30]:
1 - 0.8799700929573379

0.12002990704266214

## TF-IDF Vectorized

In [31]:
from sklearn.feature_extraction.text import TfidfVectorizer

In [32]:
content.columns

Index(['id', 'author', 'source', 'content', 'feature', 'title', 'url'], dtype='object')

In [33]:
content['is_xinhua'] = np.where(content['source'].str.contains('新华'), 1, 0)

Hint: Google: pandas, How to add new column based no condtion

In [34]:
content[content['is_xinhua'] == 1]

id            author source  \
3      89614               NaN    新华社   
17     89600               NaN    新华网   
25     89592               NaN    新华网   
27     89590               NaN    新华网   
39     89578               NaN    新华社   
51     89566               NaN    新华网   
65     89552               NaN   新华社@   
95     89522               NaN    新华社   
104    89513               NaN    新华网   
113    89504               NaN    新华社   
231    89386               NaN    新华网   
305    89312               NaN    新华社   
312    89305               NaN    新华社   
314    89303               NaN    新华社   
329    89288               NaN    新华网   
330    89287               NaN    新华网   
331    89286               NaN    新华网   
335    89282               NaN    新华网   
365    89252               NaN    新华网   
376    89241               NaN    新华网   
388    89229               NaN    新华网   
445    89172               NaN    新华网   
446    89171               NaN    新华网   
448    89169               NaN    新华网   
511    89106               NaN    新华网   
514    89103               NaN    新华网   
541    89076          王璐 PS039  新华通讯社   
543    89074               刘瑞莹    新华社   
545    89072               NaN    新华社   
552    89065               NaN    新华网   
...      ...               ...    ...   
89581     30       INACIO ROSA    新华社   
89582     29       INACIO ROSA    新华社   
89583     28       INACIO ROSA    新华社   
89584     27       INACIO ROSA    新华社   
89585     26       INACIO ROSA    新华社   
89586     25       INACIO ROSA    新华社   
89587     24      LOVELY GHOSH    新华社   
89588     23      LOVELY GHOSH    新华社   
89589     22      LOVELY GHOSH    新华社   
89590     21      LOVELY GHOSH    新华社   
89591     20      LOVELY GHOSH    新华社   
89592     19                张旌    新华社   
89593     18               NaN    新华社   
89594     17  LISELOTTE SABROE    新华社   
89595     16  LISELOTTE SABROE    新华社   
89596     15  LISELOTTE SABROE    新华社   
89597     14  LISELOTTE SABROE    新华社   
89598     13  LISELOTTE SABROE    新华社   
89599     12  LISELOTTE SABROE    新华社   
89600     11  LISELOTTE SABROE    新华社   
89601     10  LISELOTTE SABROE    新华社   
89602      9  LISELOTTE SABROE    新华社   
89603      8               郑焕松    新华社   
89604      7               郑焕松    新华社   
89605      6               郑焕松    新华社   
89606      5                邹峥    新华社   
89607      4                王衡    新华社   
89608      3                张旌    新华社   
89609      2               夏文辉    新华社   
89610      1            王子江 张寒    新华社   

                                                 content  \
3      这是 6 月 18 日 在 葡萄牙 中部 大 佩德罗 冈 地区 拍摄 的 被 森林 大火 烧...   
17     国歌 法 草案 首次 提请 最高 立法机关 审议 \r\n 点击 图片 进入 下 一页 \r...   
25     当地 时间 2017 年 6 月 17 日 ， 土耳其 马 尼萨 省 ， 在 土耳其 第 1...   
27     英国 与 欧盟 的 “ 脱欧 ” 谈判 于 19 日 正式 开始 。 然而 此时 ， 英国首...   
39     这是 6 月 18 日 在 葡萄牙 中部 大 佩德罗 冈 地区 拍摄 的 被 森林 大火 烧...   
51     戈壁 的 大漠 黄沙 曾 掩埋 了 无数 西域 古道 ， 而 如今 一条 大漠 天路 正 顽...   
65     从 百度 、 支付宝 到 App ? “ 高考 志愿 填报 ” 五花八门 \r\n 顾 女士...   
95     新华社 韩国 济州 6 月 18 日电 综述 ： 亚 投行 第二届 年会 三大 亮点 \r\...   
104    粤港澳 概念 板块 午后 走强 ， 珠海 港 、 恒基 达鑫 、 白云机场 等 个股 的 午...   
113    新华社 北京 6 月 18 日电   经 军委 领导 批准 ， 《 军营 理论 热点 怎么 ...   
231    东方 卫视 明星 旅游 体验 综艺节目 《 旅途 的 花样 》 6 月 17 日 （ 周六 ...   
305    新华社 兰州 6 月 19 日电 （ 记者 张钦 ） 记者 19 日 了解 到 ， 刚刚 出...   
312    新华社 乌鲁木齐 6 月 18 日电 （ 记者 杜刚 、 王秉阳 ） 6 月 18 日 ， ...   
314    新华社 北京 6 月 19 日电 （ 记者 熊丰 ） 记者 从 公安部 消防局 获悉 ， 1...   
329    新华社 北京 ６ 月 ２ ３ 日电 （ 记者 邢静 ） 幸福 是 人们 对 生活 最 美好 ...   
330    新华社 昆明 6 月 23 日电 （ 记者 王研 ） 近年来 ， 随着 云南 航空业 的 迅...   
331    新华社 北京 ６ 月 ２ ３ 日电 （ 记者 倪元 锦 ） ２ １ 日起 ， 京津冀 迎来 ...   
335    新华社 北京 ６ 月 ２ ３ 日电 （ 记者   倪元 锦 ） ２ １ 日起 ， 京津冀 迎...   
365    新华社 哈尔滨 6 月 23 日电 （ 记者   许正 ） 记者 23 日 从 黑龙江省 环...   
376    新华社 合肥 ６ 月 ２ ３ 日电 ?   题 ： 安徽 不断 将 跨省 流域 补偿 试点 ...   
388    新华社 重庆 ６ 月 ２ ３ 日电 ?   题 ： “ 悟空 ” 为何 跌下 共享 风口 ？...   
445    新华社 贵阳 ６ 月 ２ ３ 日电 （ 记者   李黔 渝 ） 贵州省政府 近日 下发 通知...   
446    新华社 广州 6 月 23 日电   题 ： 激发 内生 动力   走 在 脱贫 路上 — ...   
448    新华社 南京 6 月 23 日电 （ 记

In [33]:
content.head()

id      author                  source  \
0  89617         NaN  快科技@http://www.kkj.cn/   
1  89616         NaN  快科技@http://www.kkj.cn/   
2  89615         NaN  快科技@http://www.kkj.cn/   
3  89614         NaN                     新华社   
4  89613  胡淑丽_MN7479                   深圳大件事   

                                             content  \
0  此外，自本周（6月12日）起，除小米手机6等15款机型外，其余机型已暂停更新发布（含开发版/...   
1  骁龙835作为唯一通过Windows 10桌面平台认证的ARM处理器，高通强调，不会因为只考...   
2  此前的一加3T搭载的是3400mAh电池，DashCharge快充规格为5V/4A。\r\n...   
3    这是6月18日在葡萄牙中部大佩德罗冈地区拍摄的被森林大火烧毁的汽车。新华社记者张立云摄\r\n   
4  （原标题：44岁女子跑深圳约会网友被拒，暴雨中裸身奔走……）\r\n@深圳交警微博称：昨日清...   

                                             feature  \
0  {"type":"科技","site":"cnbeta","commentNum":"37"...   
1  {"type":"科技","site":"cnbeta","commentNum":"15"...   
2  {"type":"科技","site":"cnbeta","commentNum":"18"...   
3  {"type":"国际新闻","site":"环球","commentNum":"0","j...   
4  {"type":"新闻","site":"网易热门","commentNum":"978",...   

                           title  \
0           小米MIUI 9首批机型曝光：共计15款   
1     骁龙835在Windows 10上的性能表现有望改善   
2      一加手机5细节曝光：3300mAh、充半小时用1天   
3  葡森林火灾造成至少62人死亡 政府宣布进入紧急状态（组图）   
4       44岁女子约网友被拒暴雨中裸奔 交警为其披衣相随   

                                                 url is_xinhua  
0     http://www.cnbeta.com/articles/tech/623597.htm         0  
1     http://www.cnbeta.com/articles/tech/623599.htm         0  
2     http://www.cnbeta.com/articles/tech/623601.htm         0  
3  http://world.huanqiu.com/hot/2017-06/10866126....         1  
4  http://news.163.com/17/0618/00/CN617P3Q0001875...         0

## Based on Title or Based on Content? 

Answer: Just Try it

In [48]:
X_title = content['title'].values

In [49]:
X_content = content['content'].values

In [50]:
y = content['is_xinhua'].values

In [51]:
len(X_title) == len(X_content) == len(y)

True

In [52]:
y[:10]

array([0, 0, 0, 1, 0, 0, 0, 0, 0, 0])

In [17]:
from sklearn.model_selection import train_test_split

## Get X_title tfidf vector

In [40]:
from sklearn.feature_extraction.text import TfidfVectorizer

In [46]:
vectorizer = TfidfVectorizer(max_features=30000)

In [53]:
X_title = vectorizer.fit_transform(X_title)

In [54]:
X_content = vectorizer.fit_transform(X_content)

In [55]:
X_content.shape

(89611, 30000)

In [23]:
X_t_train, X_t_test, y_train, y_test = train_test_split(X_title, y, test_size=0.25, random_state=0)

NameError: name 'train_test_split' is not defined

In [79]:
X_c_train, X_c_test, y_c_train, y_c_test = train_test_split(X_content, y, test_size=0.25, random_state=0)

In [78]:
X_t_train.shape

(67208, 70108)

In [94]:
indices = np.arange(len(y))
np.random.shuffle(indices)
split_test = 0.25

In [101]:
train_indices = indices[: int(len(indices)*split_test)]

In [102]:
test_indices = indices[int(len(indices) * split_test):]

In [131]:
X_t_train, X_t_test, y_train, y_test = (
    X_title[train_indices], X_title[test_indices], 
    y[train_indices], y[test_indices])

In [129]:
from sklearn.naive_bayes import GaussianNB

In [133]:
clf = GaussianNB()

In [134]:
from sklearn import datasets

In [136]:
clf.fit(X_t_train.toarray(), y_train)

GaussianNB(priors=None)

In [199]:
from sklearn.metrics import confusion_matrix

In [201]:
def get_performance(clf, X_, y_, sample=1000):
    random_indices = np.random.choice(np.arange(len(y_)), sample)
    sub_x = X_[random_indices]
    sub_y = y_[random_indices]
    y_hat = clf.predict(sub_x.toarray())
    print('percision is: {}'.format(precision_score(sub_y, y_hat)))
    print('recall is: {}'.format(recall_score(sub_y, y_hat)))
    print('roc_auc is: {}'.format(roc_auc_score(sub_y, y_hat)))
    print('confusion matrix: \n{}'.format(confusion_matrix(sub_y, y_hat, labels=[0, 1])))

In [158]:
sample = 5000
random_train_indices = np.random.choice(np.arange(len(y_train)), sample)

In [159]:
sub_x_train = X_t_train[random_train_indices]

In [160]:
sub_y_train = y_train[random_train_indices]

In [161]:
sub_predict = clf.predict(sub_x_train.toarray())

In [184]:
from sklearn.metrics import precision_score

In [152]:
from sklearn.metrics import recall_score

In [166]:
from sklearn.metrics import roc_auc_score

In [186]:
get_performance(clf, X_=X_t_train, y_=y_train, sample=1000)

percision is: 1.0
recall is: 1.0
roc_auc is: 1.0


By definition a confusion matrix $ C  $ is such that $ C_{i, j} $ is equal to the number of observations known to be in group $ i $  but predicted to be in group $ j $.

Thus in binary classification, the count of true negatives is $ C_{0,0} $, false negatives is $ C_{1,0} $ , true positives is $ C_{1,1} $ and false positives is $ C_{0,1} $.

In [202]:
get_performance(clf, X_=X_t_test, y_=y_test, sample=3000)

percision is: 0.9948330683624801
recall is: 0.9364010475121586
roc_auc is: 0.9483228479151007
confusion matrix: 
[[ 314   13]
 [ 170 2503]]


314 is the number that is group 0, and classified to group 0, 13 is the number that is group 0, but classified to 1; 170 is the number that is group 1, but is classified to group 0, 2503 is group 1 and classified to group 1.

In [163]:
accuracy_score(sub_y_train, sub_predict)

1.0

In [164]:
recall_score(sub_y_train, sub_predict)

1.0

In [167]:
roc_auc_score(sub_y_train, sub_predict)

1.0

In [168]:
test_sample = 5000
random_test_indices = np.random.choice(np.arange(len(y_test)), test_sample)

In [169]:
sub_x_test = X_t_test[random_test_indices]

In [171]:
sub_y_test = y_test[random_test_indices]

In [174]:
sub_test_predict = clf.predict(sub_x_test.toarray())

In [175]:
accuracy_score(sub_y_test, sub_test_predict)

0.9344

In [176]:
recall_score(sub_y_test, sub_test_predict)

0.9282758620689655

In [177]:
roc_auc_score(sub_y_test, sub_test_predict)

0.9518302387267905

## If we add some prior how performance? 

In [196]:
clf_p = GaussianNB(priors=[0.5, 0.5])

In [197]:
clf_p.fit(X_t_train.toarray()[:3000], y_train[:3000])

GaussianNB(priors=[0.5, 0.5])

In [195]:
get_performance(clf_p, X_=X_t_train, y_=y_train, sample=1000) # old

percision is: 0.996328029375765
recall is: 0.9187358916478555
roc_auc is: 0.9462100510870856


In [198]:
get_performance(clf_p, X_=X_t_train, y_=y_train, sample=1000) # new 0.5 0.5

percision is: 0.9962216624685138
recall is: 0.9197674418604651
roc_auc is: 0.9491694352159469


### performance of classification with prior

In [190]:
get_performance(clf_p, X_=X_t_test, y_=y_test, sample=1000)

percision is: 0.9975874547647768
recall is: 0.9229910714285714
roc_auc is: 0.951880151098901


In [203]:
from sklearn.tree import DecisionTreeClassifier

In [254]:
clf_tree = DecisionTreeClassifier(class_weight={0:5, 1:4}, criterion='entropy', max_features=5000)

In [209]:
len(y_train)

22402

In [256]:
clf_tree.fit(X_t_train[:2200], y_train[:2200])

DecisionTreeClassifier(class_weight={0: 5, 1: 4}, criterion='entropy',
            max_depth=None, max_features=5000, max_leaf_nodes=None,
            min_impurity_decrease=0.0, min_impurity_split=None,
            min_samples_leaf=1, min_samples_split=2,
            min_weight_fraction_leaf=0.0, presort=False, random_state=None,
            splitter='best')

In [211]:
get_performance(clf_tree, X_=X_t_train, y_=y_train, sample=1000)

percision is: 0.9987228607918263
recall is: 0.8916761687571265
roc_auc is: 0.9417730437281566
confusion matrix: 
[[122   1]
 [ 95 782]]


In [257]:
get_performance(clf_tree, X_=X_t_train, y_=y_train, sample=1000)

percision is: 0.9987421383647799
recall is: 0.8911335578002245
roc_auc is: 0.9409796229368096
confusion matrix: 
[[108   1]
 [ 97 794]]


In [212]:
get_performance(clf_tree, X_=X_t_test, y_=y_test, sample=1000)

percision is: 0.9987029831387808
recall is: 0.8901734104046243
roc_auc is: 0.9413830014986084
confusion matrix: 
[[134   1]
 [ 95 770]]


### Find the most similar 

In [56]:
mini_X_title = X_title[:1000]

In [25]:
del X_title

In [26]:
del X_content

In [57]:
mini_X_content = X_content[:1000]

In [62]:
mini_X_title_ndarray = mini_X_title.toarray()
mini_X_content_ndarray = mini_X_content.toarray()

In [59]:
del mini_X_content_ndarray

In [60]:
del mini_X_title_ndarray

if we change the [0, 0, a, 0, 0, 0] = [False, False, True, False, False, False]

In [63]:
X_content_ndarray_one_zero = mini_X_content_ndarray.astype(bool).astype(int)

In [65]:
X_content_ndarray_transpose = np.transpose(X_content_ndarray_one_zero)

In [66]:
X_words_article_id_map = X_content_ndarray_transpose

In [71]:
np.where(X_content[0].toarray())[1]

array([   39,    78,   768,  2465,  2516,  4244,  4828,  5057,  5128,
        7249,  7445,  7715, 11427, 11846, 11893, 12685, 13041, 13082,
       13812, 13988, 16188, 16482, 16548, 16549, 16667, 16875, 16971,
       17898, 20132, 20580, 21448, 21497, 21500, 22354, 22545, 26893,
       27012, 28659])

In [74]:
id_word_map = {i: w for w, i in vectorizer.vocabulary_.items()}

In [77]:
for i in np.where(X_content[0].toarray())[1]: 
    print(id_word_map[i])

12
15
一年
主要
之中
信息
全部
关于
其余
去年
发布
可以
小米
工作
工程师
开发
当然
影响
我们
手机
时候
暂停
更新
更新换代
有人
本周
机型
此外
猜测
用到
研发
确保
确切
等待
精力
还是
进行
集中


In [72]:
vectorizer.vocabulary_

{'此外': 17898,
 '本周': 16875,
 '12': 39,
 '小米': 11427,
 '手机': 13988,
 '15': 78,
 '机型': 16971,
 '其余': 5128,
 '暂停': 16482,
 '更新': 16548,
 '发布': 7445,
 '开发': 12685,
 '影响': 13082,
 '确保': 21497,
 '工程师': 11893,
 '可以': 7715,
 '集中': 28659,
 '全部': 4828,
 '精力': 22545,
 '进行': 27012,
 '工作': 11846,
 '有人': 16667,
 '猜测': 20132,
 '主要': 2465,
 '用到': 20580,
 '研发': 21448,
 '之中': 2516,
 '去年': 7249,
 '一年': 768,
 '时候': 16188,
 '更新换代': 16549,
 '当然': 13041,
 '关于': 5057,
 '确切': 21500,
 '信息': 4244,
 '我们': 13812,
 '还是': 26893,
 '等待': 22354,
 '骁龙': 29556,
 '作为': 3963,
 '唯一': 8447,
 '通过': 27342,
 'windows': 621,
 '10': 14,
 '桌面': 17574,
 '平台': 12281,
 '认证': 25341,
 '处理器': 9524,
 '高通': 29725,
 '强调': 12987,
 '不会': 1403,
 '因为': 8690,
 '考虑': 23538,
 '性能': 13418,
 '核心': 17493,
 '相反': 21167,
 '他们': 3413,
 '联手': 23656,
 '微软': 13224,
 '找到': 14206,
 '一种': 883,
 '适合': 27203,
 '兼顾': 5183,
 '完美': 10874,
 '方案': 15917,
 '报道': 14418,
 '已经': 11974,
 '拿到': 14634,
 '一些': 650,
 '以便': 3461,
 '更好': 16541,
 '理解': 20416,
 '架构': 17350,
 '资

In [78]:
from functools import reduce

In [79]:
import operator as op

In [84]:
def prodIntersection(sets):
    return reduce(op.and_, sets)

In [85]:
prodIntersection([{1, 2, 3}, {1, 2}])

{1, 2}

In [167]:
article_one_words_id = np.where(X_content[0].toarray()[0])[0]

In [174]:
np.sort(X_content[0].toarray()[0])[-20:][::-1]

array([0.40030771, 0.3566469 , 0.22755048, 0.22338446, 0.21054757,
       0.20015385, 0.19571384, 0.18398833, 0.16527315, 0.16468233,
       0.16180044, 0.1610011 , 0.15523609, 0.15457139, 0.14737251,
       0.146817  , 0.13426477, 0.13027367, 0.12815933, 0.12318049])

In [155]:
np.argmax(X_content[0].toarray(), out=)

16971

In [182]:
def get_article_by_one_input(article_id, match_num=1.):
    article_wors_id = np.where(X_content[article_id].toarray())[1]
    if isinstance(match_num, float):
        num = int(len(article_wors_id) * match_num)
        ids = np.random.choice(article_wors_id, num)
    elif isinstance(match_num, int):
        num = match_num
        ids = np.argsort(X_content[article_id].toarray()[0])[-num:][::-1]
    return prodIntersection(get_articles_by_word_id(i) for i in ids)

In [110]:
prodIntersection(get_articles_by_word_id(i) for i in article_one_words_id)

{0}

In [183]:
get_article_by_one_input(article_id=3, match_num=10)

{3, 39}

In [131]:
content.iloc[74]

id                                                       89543
author                                                     NaN
source                                                   中国经济网
content      免责 声明 ： 本文 仅 代表 作者 个人观点 ， 与 环球网 无关 。 其 原创性 以及 ...
feature      {"type":"国内新闻","site":"环球","commentNum":"0","j...
title                          【 图解 】 香港 各界 政要 热议 “ 一带 一路 ” 倡议
url          http://china.huanqiu.com/hot/2017-06/10866428....
is_xinhua                                                    0
Name: 74, dtype: object

In [134]:
# article: 423 match set: {408, 411, 423}
content.iloc[408]

id                                                       89209
author                                                     NaN
source                                                   中国新闻网
content      中新社 江苏 盐城 6 月 23 日电   ( 记者   于 从文 ) 粉墙黛瓦 ， 花木扶...
feature      {"type":"社会滚动","site":"参考消息","commentNum":"0",...
title                 盐城 “ 6 · 23 ” 风灾 一年 后 5164 户 受灾 民众 迁入 新居
url          http://www.cankaoxiaoxi.com/society/20170623/2...
is_xinhua                                                    0
Name: 408, dtype: object

In [135]:
content.iloc[411]

id                                                       89206
author                                                     NaN
source                                                   中国新闻网
content      中新社 江苏 盐城 6 月 23 日电   ( 记者   于 从文 ) 粉墙黛瓦 ， 花木扶...
feature      {"type":"社会滚动供wifi","site":"参考消息","commentNum"...
title                 盐城 “ 6 · 23 ” 风灾 一年 后 5164 户 受灾 民众 迁入 新居
url          http://www.cankaoxiaoxi.com/society/rollwifi/2...
is_xinhua                                                    0
Name: 411, dtype: object

In [137]:
content.iloc[423]

id                                                       89194
author                                                     NaN
source                                                   中国新闻网
content      中新社 江苏 盐城 6 月 23 日电   ( 记者   于 从文 ) 粉墙黛瓦 ， 花木扶...
feature      {"type":"社会观察","site":"参考消息","commentNum":"0",...
title                 盐城 “ 6 · 23 ” 风灾 一年 后 5164 户 受灾 民众 迁入 新居
url          http://www.cankaoxiaoxi.com/china/20170623/214...
is_xinhua                                                    0
Name: 423, dtype: object

In [139]:
content.iloc[10]['content']

'\u3000 \u3000 证券时报 网 （ www . stcn . com ） 06 月 23 日讯 \r\n \u3000 \u3000 据 上证 报道 ， 6 月初 以来 ， 创业板 指数 明显 强于 上证综指 ， 一时间 关于 蓝筹 行情 将 向 成长 股行情 切换 的 预期 再起 。 但 多位 私募 人士 表示 ， 蓝筹股 与 成长 股 这 两年 并非 绝对 的 此消彼长 。 本轮 蓝筹股 上涨 行情 的 主要 推动力 就是 业绩 成长 ， 而 业绩 与 估值 匹配 才 是 王道 。 上述 私募 人士 依然 看好 成长性 突出 的 蓝筹股 。 \r\n'

In [140]:
# 29, 131, 263, 135, 138, 267, 12, 140, 141, 275, 150, 281, 27, 30, 31, 286, 291, 41, 298, 48, 305, 53, 55, 312, 57, 314, 63, 66, 71, 72, 74, 78, 80, 85, 90, 218, 221, 95, 225, 227, 228, 102, 110, 113, 242, 245, 118, 247, 122, 125, 127

In [148]:
def print_content(ids):
    for i in ids:
        print('id == {} from == {} content = {}'.format(i, content.iloc[i]['source'], content.iloc[i]['content'][:100]))

In [187]:
print_content({413, 2366, 159})

id == 413 from == 北京青年报@ content = 银监会 等 部门 要求 网贷 平台 一律 暂停 校园 贷 业务 ， 并 明确 退出 时间表 。 银监会 、 教育部 、 人力资源 社会保障部 近日 下发 通知 ， 要求 未经 银行业 监管部门 批准 
id == 2366 from == 每日经济新闻 content = 近年来 ， 有关 校园 “ 裸贷 ” 、 “ 暴力 催收 ” 等 事件 频频 曝光 。 
 一些 鉴别能力 不高 ， 社会 经验不足 的 大学生 ， 或 出于 渴望 消费 、 满足 物欲 ， 或是 
id == 159 from == 每日经济新闻@ content = 《 通知 》 称 ， 商业银行 和 政策性 银行 在 风险 可控 的 前提 下 ， 有 针对性 地 开发 高校 助学 、 培训 、 消费 、 创业 等 金融 产品 ， 而 未经 银行业 监管部门 批准


In [186]:
test_mode = True
for i, a in enumerate(range(len(content))):
    if test_mode and i > 3000: break
    try:
        match_articles = get_article_by_one_input(article_id=a, match_num=20)
    except ValueError: pass
    
    match_articles.add(a)
    if len(match_articles) > 1: 
        print('article: {} match set: {}'.format(a, match_articles))

article: 17 match set: {17, 653}
article: 65 match set: {65, 431}
article: 74 match set: {129, 131, 263, 135, 138, 267, 12, 140, 141, 275, 150, 281, 27, 30, 286, 31, 291, 41, 298, 48, 305, 53, 55, 312, 57, 314, 63, 66, 71, 72, 74, 78, 80, 85, 90, 218, 221, 95, 225, 227, 228, 102, 110, 113, 242, 245, 118, 247, 122, 125, 127}
article: 78 match set: {314, 78}
article: 131 match set: {298, 131}
article: 159 match set: {413, 159}
article: 236 match set: {948, 236}
article: 262 match set: {112, 262}
article: 296 match set: {296, 790}
article: 298 match set: {298, 131}
article: 314 match set: {314, 78}
article: 331 match set: {331, 335}
article: 335 match set: {331, 335}
article: 386 match set: {386, 394}
article: 394 match set: {386, 394}
article: 406 match set: {426, 406, 407}
article: 407 match set: {426, 406, 407}
article: 408 match set: {408, 411, 423}
article: 411 match set: {408, 411, 423}
article: 417 match set: {417, 419, 956}
article: 419 match set: {417, 419, 956}
article: 423 matc

In [122]:
content.iloc[3]

id                                                       89614
author                                                     NaN
source                                                     新华社
content      这是 6 月 18 日 在 葡萄牙 中部 大 佩德罗 冈 地区 拍摄 的 被 森林 大火 烧...
feature      {"type":"国际新闻","site":"环球","commentNum":"0","j...
title             葡 森林 火灾 造成 至少 62 人 死亡   政府 宣布 进入 紧急状态 （ 组图 ）
url          http://world.huanqiu.com/hot/2017-06/10866126....
is_xinhua                                                    1
Name: 3, dtype: object

In [123]:
content.iloc[39]

id                                                       89578
author                                                     NaN
source                                                     新华社
content      这是 6 月 18 日 在 葡萄牙 中部 大 佩德罗 冈 地区 拍摄 的 被 森林 大火 烧...
feature      {"type":"国际新闻","site":"环球","commentNum":"0","j...
title                    葡 森林 火灾 造成 至少 62 人 死亡   政府 宣布 进入 紧急状态
url          http://world.huanqiu.com/hot/2017-06/10866229....
is_xinhua                                                    1
Name: 39, dtype: object

In [94]:
def get_articles_by_word_id(word_id):
    return set(np.where(X_words_article_id_map[word_id])[0])

In [96]:
set(np.where(X_words_article_id_map[39])[0])

{0,
 6,
 7,
 8,
 19,
 27,
 29,
 43,
 46,
 47,
 63,
 68,
 70,
 71,
 84,
 93,
 95,
 97,
 112,
 124,
 138,
 144,
 147,
 150,
 153,
 158,
 159,
 170,
 172,
 185,
 186,
 190,
 191,
 195,
 205,
 209,
 224,
 237,
 242,
 243,
 245,
 248,
 257,
 259,
 262,
 267,
 268,
 269,
 273,
 275,
 285,
 296,
 299,
 323,
 325,
 327,
 333,
 346,
 349,
 366,
 384,
 385,
 387,
 389,
 401,
 404,
 414,
 443,
 454,
 456,
 461,
 466,
 474,
 479,
 487,
 498,
 502,
 504,
 511,
 512,
 522,
 527,
 529,
 530,
 543,
 547,
 553,
 555,
 562,
 566,
 568,
 575,
 577,
 578,
 582,
 600,
 604,
 606,
 607,
 616,
 617,
 629,
 638,
 640,
 647,
 650,
 656,
 661,
 685,
 691,
 700,
 717,
 725,
 726,
 728,
 733,
 735,
 743,
 754,
 761,
 762,
 769,
 771,
 782,
 783,
 790,
 798,
 801,
 803,
 804,
 805,
 806,
 809,
 811,
 812,
 814,
 817,
 818,
 822,
 823,
 824,
 827,
 828,
 829,
 830,
 833,
 834,
 835,
 836,
 837,
 840,
 844,
 845,
 846,
 848,
 864,
 875,
 879,
 885,
 914,
 920,
 927,
 929,
 932,
 933,
 944,
 956,
 962,
 972,
 973,
 9